price-feeder/oracle/provider/crypto.go

```diff
@@ -0,0 +1,536 @@
+			key,
+		)
+	}
+
+	return candles, nil
+}
+
+func (p *CryptoProvider) messageReceived(messageType int, bz []byte, reconnectTicker *time.Ticker) {
+	if messageType != websocket.TextMessage {
+		return
+	}
+
+	var (
+		heartbeatResp CryptoHeartbeatResponse
+		tickerResp    CryptoTickerResponse
+		tickerErr     error
+		candleResp    CryptoCandleResponse
+		candleErr     error
+	)
+
+	// sometimes the message received is not a ticker or a candle response.
+	_ = json.Unmarshal(bz, &heartbeatResp)
+	if heartbeatResp.Method == cryptoHeartbeatMethod {
+		p.pong(bz, reconnectTicker)
+		return
+	}
+
+	tickerErr = json.Unmarshal(bz, &tickerResp)
+	if tickerResp.Result.Channel == cryptoTickerChannel {
+		for _, tickerPair := range tickerResp.Result.Data {
+			p.setTickerPair(
+				tickerResp.Result.InstrumentName,
+				tickerPair,
+			)
+			telemetryWebsocketMessage(ProviderCrypto, MessageTypeTicker)
+		}
+		return
+	}
+
+	candleErr = json.Unmarshal(bz, &candleResp)
+	if candleResp.Result.Channel == cryptoCandleChannel {
+		for _, candlePair := range candleResp.Result.Data {
+			p.setCandlePair(
+				candleResp.Result.InstrumentName,
+				candlePair,
+			)
+			telemetryWebsocketMessage(ProviderCrypto, MessageTypeCandle)
+		}
+		return
+	}
+
+	p.logger.Error().
+		Int("length", len(bz)).
+		AnErr("ticker", tickerErr).
+		AnErr("candle", candleErr).
+		Msg("Error on receive message")
+}
+
+// pong return a heartbeat message when a "ping" is received and reset the
+// recconnect ticker because the connection is alive. After connected to crypto.com's
+// Websocket server, the server will send heartbeat periodically (30s interval).
+// When client receives an heartbeat message, it must respond back with the
+// public/respond-heartbeat method, using the same matching id,
+// within 5 seconds, or the connection will break.
+func (p *CryptoProvider) pong(bz []byte, reconnectTicker *time.Ticker) {
+	reconnectTicker.Reset(cryptoReconnectTime)
+	var (
+		heartbeatResp CryptoHeartbeatResponse
+		heartbeatReq  CryptoHeartbeatRequest
+	)
+
+	if err := json.Unmarshal(bz, &heartbeatResp); err != nil {
+		p.logger.Err(err).Msg("could not unmarshal heartbeat")
+		return
+	}
+
+	heartbeatReq.ID = heartbeatResp.ID
+	heartbeatReq.Method = "public/respond-heartbeat"
+
+	if err := p.wsClient.WriteJSON(heartbeatReq); err != nil {
+		p.logger.Err(err).Msg("could not send pong message back")
+	}
+}
+
+// ping to check websocket connection
+func (p *CryptoProvider) ping() error {
+	return p.wsClient.WriteMessage(websocket.PingMessage, ping)
+}
+
+func (p *CryptoProvider) setTickerPair(symbol string, tickerPair CryptoTicker) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	price, err := coin.NewDecFromFloat(tickerPair.LatestTrade)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker price")
+	}
+	volume, err := coin.NewDecFromFloat(tickerPair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker volume")
+	}
```

Comment:
```suggestion
	price, err := coin.NewDecFromFloat(tickerPair.LatestTrade)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker price")
		return
	}
	volume, err := coin.NewDecFromFloat(tickerPair.Volume)
	if err != nil {
		p.logger.Warn().Err(err).Msg("crypto: failed to parse ticker volume")
		return
	}
```

We should not set if there is a problem with parsing values